shared/util_linux.go

```diff
@@ -384,3 +386,90 @@ func SetSize(fd int, width int, height int) (err error) {
 	}
 	return nil
 }
+
+// This uses ssize_t llistxattr(const char *path, char *list, size_t size); to
+// handle symbolic links (should it in the future be possible to set extended
+// attributed on symlinks): If path is a symbolic link the extended attributes
+// associated with the link itself are retrieved.
+func llistxattr(path string, list []byte) (sz int, err error) {
+	var _p0 *byte
+	_p0, err = unix.BytePtrFromString(path)
+	if err != nil {
+		return
+	}
+	var _p1 unsafe.Pointer
+	if len(list) > 0 {
+		_p1 = unsafe.Pointer(&list[0])
+	} else {
+		_p1 = unsafe.Pointer(nil)
+	}
+	r0, _, e1 := unix.Syscall(unix.SYS_LLISTXATTR, uintptr(unsafe.Pointer(_p0)), uintptr(_p1), uintptr(len(list)))
+	sz = int(r0)
+	if e1 != 0 {
+		err = e1
+	}
+	return
+}
+
+// GetAllXattr retrieves all extended attributes associated with a file,
+// directory or symbolic link.
+func GetAllXattr(path string) (xattrs map[string]string, err error) {
+	e1 := errors.New("Extended attributes changed during retrieval.")
+
+	pre, err := llistxattr(path, nil)
+	if err != nil || pre < 0 {
+		return nil, err
+	}
+	if pre == 0 {
+		return nil, nil
+	}
+
+	dest := make([]byte, pre)
+
+	post, err := llistxattr(path, dest)
+	if err != nil || post < 0 {
+		return nil, err
+	}
+	if post != pre {
+		return nil, e1
+	}
+
+	split := strings.Split(string(dest), "\x00")
+	if split == nil {
+		return nil, errors.New("No valid extended attribute key found.")
```

Comment:
Everywhere else we use `fmt.Errorf`; I'm not sure there's really a difference, but in case someone goes through and does better error handling in the future, it might be good to be consistent here.